handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,221 @@
+                        if (removed) {
+                            return;
+                        }
+                        writeFinishTime = System.nanoTime();
+                        //init readFinishTime to writeFinishTime - readFinishTime > 0 be true if we don't get response
+                        readFinishTime = writeFinishTime - 1;
+                        if (timeoutNanos > 0) {
+                            timeout = ctx.executor().schedule(readTimeoutTask, timeoutNanos, TimeUnit.NANOSECONDS);
+                        }
+                    }
+                }
+            };
+        }
+
+        if (readTimeoutTask == null) {
+            readTimeoutTask = new ReadTimeoutTask(ctx);
+        }
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        readFinishTime = System.nanoTime();
+        if (timeoutShouldBeCancelled(ctx, msg)) {
+            disable();
+        }
+        if (suppressReadAfterTimeout && timedOut) {
+            ctx.fireUserEventTriggered(ReadSuppressed.INSTANCE);
+            ReferenceCountUtil.release(msg);
+        } else {
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelOnWriteListener));
+    }
+
+    /**
+     * This method is called after receiving each {@link #channelRead} and should define if this read should trigger
+     * timeout cancelling. Default implementation returns {@code true}.
+     * @param ctx {@link ChannelHandlerContext} of related {@link #channelRead} invocation
+     * @param msg message of related {@link #channelRead} invocation
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    protected boolean timeoutShouldBeCancelled(@SuppressWarnings("unused") ChannelHandlerContext ctx,
+                                               @SuppressWarnings("unused") Object msg) {
+        return true;
+    }
+
+    private void destroy() {
+        removed = true;
+        disable();
+    }
+
+    private void disable() {
+        if (timeout != null) {
+            timeout.cancel(false);
+            timeout = null;
+        }
+    }
+
+    /**
+     * Is called when a read timeout was detected.
+     */
+    protected void readTimedOut(ChannelHandlerContext ctx) throws Exception {
+        timedOut = true;
+        ctx.fireExceptionCaught(ReadTimeoutException.INSTANCE);
+        if (closeOnTimeout) {
+            ctx.close();
+        }
+    }
+
+    public static final class ReadSuppressed {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)